driver-core/src/main/com/mongodb/client/model/expressions/DocumentExpression.java

```diff
@@ -16,10 +16,77 @@
 
 package com.mongodb.client.model.expressions;
 
+import org.bson.conversions.Bson;
+
+import java.time.Instant;
+
+import static com.mongodb.client.model.expressions.Expressions.of;
+
 /**
  * Expresses a document value. A document is an ordered set of fields, where the
  * key is a string value, mapping to a value of any other expression type.
  */
 public interface DocumentExpression extends Expression {
```

Comment:
Both setField and getField allow the field name itself to be an expression.  That API should support that as well.

Reply:
This is supported for Map<T> (arbitrary number of string keys to values all of the same type), but is not supported for Documents, which represent records (fixed number of string-named fields to values of different types). Neither one extends the other, because:

1. A record cannot have a `set(T)` method since this would allow fields to be set via a broader type than they actually are.
2. A `map<StringEx>` cannot have a method `getBoolean(field)` because this would allow string values to be pulled out as booleans.

Any situation that requires this is almost certainly more correctly handled via a `MapExpression<Expression>`, rather than a document (record).

Reply:
What is a `MapExpression`?  I don't see that anywhere.

Reply:
It's in https://github.com/mongodb/mongo-java-driver/pull/1054

Reply:
Can you show me what the code would look like?  I don't understand how it works in practice, e.g. say I have a document like:

```
{
   s1 : "str1",
   s2: "str2",
   s3 : "str3",
   i: 1
}
```

How would I do something like set one of the `s` fields where the suffix of the field name is the value of the `i` field?

Reply:
Deferring at least to map expression PR